Adding support for Word 2013 and 2016

ManoShu opened the pull request:
Added the new namespaces from 2013 and 2016, not exactly sure if the tags on W15 and W16se, but it solved my problem as described on #102

msftclas commented:
Hi __@ManoShu__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

msftgits commented:
Hi, I am closing and re-opening this PR to bump the CLA bot. Sorry for the inconvenience!

msftclas commented:
Hi __@ManoShu__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

ThomasBarnekow commented on OpenXmlPowerTools/PtOpenXmlUtil.cs:
```diff
@@ -5359,6 +5359,222 @@ public static class W14
+    public static class W15
+    {
+        public static XNamespace w15 =
+            "http://schemas.microsoft.com/office/word/2012/wordml";
```
Could you turn `public static` into `public static readonly`?

ThomasBarnekow commented on OpenXmlPowerTools/PtOpenXmlUtil.cs:
```diff
@@ -5359,6 +5359,222 @@ public static class W14
+
+    public static class W16se
+    {
+        public static XNamespace w16se =
```
As for the `W15` class, could you turn `public static` into `public static readonly`? In my PR, I have changed all other fields into `public static readonly` fields.

ManoShu commented:
I added the readonly attribute to the fields. Is that all?

ThomasBarnekow commented:
Hi @ManoShu, sorry for the late reply. I would not have anything else. Let's see what @EricWhiteDev has to say.

ThomasBarnekow commented:
@ManoShu and @EricWhiteDev, looking at this PR again, I saw that a few namespaces that you would find in Microsoft Word 2016 documents are not included in the lists of namespaces in `MarkupSimplifier.cs` and `DocumentBuilder.cs` as amended by this PR (and certainly not before).

Here's the `w:document` tag as produced by Word 2016 with markup added by myself to point out the differences that I saw:
```
<w:document 
  xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" 
  xmlns:cx="http://schemas.microsoft.com/office/drawing/2014/chartex"          <== MISSING
  xmlns:cx1="http://schemas.microsoft.com/office/drawing/2015/9/8/chartex"     <== MISSING
  xmlns:cx2="http://schemas.microsoft.com/office/drawing/2015/10/21/chartex"   <== MISSING
  xmlns:cx3="http://schemas.microsoft.com/office/drawing/2016/5/9/chartex"     <== MISSING
  xmlns:cx4="http://schemas.microsoft.com/office/drawing/2016/5/10/chartex"    <== MISSING
  xmlns:cx5="http://schemas.microsoft.com/office/drawing/2016/5/11/chartex"    <== MISSING
  xmlns:cx6="http://schemas.microsoft.com/office/drawing/2016/5/12/chartex"    <== MISSING
  xmlns:cx7="http://schemas.microsoft.com/office/drawing/2016/5/13/chartex"    <== MISSING
  xmlns:cx8="http://schemas.microsoft.com/office/drawing/2016/5/14/chartex"    <== MISSING
  xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" 
  xmlns:aink="http://schemas.microsoft.com/office/drawing/2016/ink" 
  xmlns:am3d="http://schemas.microsoft.com/office/drawing/2017/model3d" 
  xmlns:o="urn:schemas-microsoft-com:office:office" 
  xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" 
  xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" 
  xmlns:v="urn:schemas-microsoft-com:vml" 
  xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" 
  xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" 
  xmlns:w10="urn:schemas-microsoft-com:office:word" 
  xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" 
  xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" 
  xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml" 
  xmlns:w16cid="http://schemas.microsoft.com/office/word/2016/wordml/cid"      <== MISSING
  xmlns:w16se="http://schemas.microsoft.com/office/word/2015/wordml/symex" 
  xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" 
  xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" 
  xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" 
  xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" 
  mc:Ignorable="w14 w15 w16se w16cid wp14">                                    <== DIFFERENT
```
For example, if I understand the code in MarkupSimplifier correctly, it uses the namespaces defined there  for the `w:document` element if it is asked to normalize the XML. This would mean that Word 2016 documents would have the missing namespaces removed from the `w:document` element. If this is a problem, this means that this PR should be further amended to include the missing namespaces. Further, it means that this PR is also important, because as it currently stands, none of the namespaces added with this PR are currently taken into account.

ManoShu commented:
I added the namespaces which were pointed out. 
Also, I just found out about this definition page for the DOCX schemas [here](https://msdn.microsoft.com/en-us/library/dd952656(v=office.12).aspx); It is necessary to add those elements as well?

EricWhiteDev commented:
@ManoShu I don't have any information regarding those namespaces.  What are the namespace prefixes for those namespaces - can you research?

ThomasBarnekow commented:
@ManoShu Thanks for updating the pull request.

I would ask @EricWhiteDev to look into this, because some changes relate to classes (e.g., WmlComparer) that he currently actively develops. However, from my point of view, there would be at least two things that we should do (one of which will have to be synchronized with @EricWhiteDev's activities.

Firstly, @ManoShu should rebase this pull request on vNext. We currently have conflicts that need to be resolved. Those are coming from the recent upgrade to v2.7.2 of the Open XML SDK. And as you have not added any files to any project, I would not expect to see any changes to the .csproj files once you've rebased. On the other hand, that creates the question of whether or not you should have added unit tests and/or test documents ;-). 

Secondly, we currently have three classes (DocumentBuilder, MarkupSimplifier, and WmlComparer) in which arrays like this are declared:
```
        private static XAttribute[] NamespaceAttributes =
        {
            new XAttribute(XNamespace.Xmlns + "wpc", WPC.wpc),
            new XAttribute(XNamespace.Xmlns + "mc", MC.mc),
            new XAttribute(XNamespace.Xmlns + "o", O.o),
            [...]
        };
```
I understand that @ManoShu has just worked with what was in the code base. However, if we amend it now, it makes sense to extract this and put it in some central place (otherwise the canonical unit test would be to ensure these arrays are all equal ;-). This obviously needs to be synchronized with @EricWhiteDev's activities.

EricWhiteDev commented:
@ManoShu @ThomasBarnekow At the exact moment, I have no changes that need to be merged into vNext.  It is up to date with all development that I had been doing in branches in other repos.  Therefore this is an ideal time to update this.  It would make sense to put this array in the PtOpenXmlUtil.cs module, and reference it from other modules.

ThomasBarnekow commented:
@EricWhiteDev If you've run out of changes that are ready to be merged, there would be my PR, which fixes the `RevisionProcessor` ;-). However, I think I'd have to rebase that as well. So, depending on how quickly @ManoShu can rebase his PR and refactor as just discussed, you could merge his PR first.

EricWhiteDev commented:
@ThomasBarnekow OK, I'm happy to do either, as you recommend.  I will probably want to do some extensive tests on the RevisionProcessor - I have a large number of tests in a private repo (must be private because contains documents that are proprietary to a customer), so will want to run those tests.  But it makes sense to:
- If @ManoShu can rebase and refactor, then merge that PR
- You rebase the RevisionProcessor PR
- I do my extensive tests
- Merge the RevisionProcessor PR

EricWhiteDev commented:
OK, I will run my tests.

EricWhiteDev commented:
@ManoShu We have already tested and merged the RevisionProcessor PR - so if you can refactor, and rebase, we can merge this PR.  Can you move that array of namespaces into PtOpenXmlUtil.cs?

ManoShu commented:
@EricWhiteDev OK, rebase was done. here is what i changed:

- Concentrated the XAttribute arrays (a common list and a Power Tools exclusive) on a new NamespaceAttributeUtil class on PtOpenXmlUtil.cs;
- Replaced any string-based namespace declaration to XElement manipulation;
- Kept the array declarations on the previous spots, now just referencing the NamespaceAttributeUtil lists;

Sorry for not adding additional tests and documents, all the 2016 DOCX files I have are from customers.

ThomasBarnekow commented:
@ManoShu Not sure how exactly you rebased, but I am seeing 92 commits and 469 changed files in this pull request. We definitely can't merge it in this state.

ManoShu commented:
@ThomasBarnekow Sorry, I have done something wrong about this rebase, i used [this](https://stackoverflow.com/a/7244456) list of commands, but instead of using git merge, I used git rebase.
I will rollback those commits and make it the proper way, and again sorry, my usual set of git commands used is very basic.

ThomasBarnekow commented:
@ManoShu This is always easier if you have made your changes on a feature branch. You used vNext directly, which makes it a bit harder to do right.

Normally, if you use a feature branch, say `support-word-2016`, you would simply make sure your origin vNext is in sync with upstream vNext and then rebase your feature branch onto vNext. For example:
```
git checkout vNext
git pull upstream vNext
git rebase vNext support-word-2016
git push --force origin
```
If you are working on vNext directly, which is not the recommended practice, the following might work. But note that I have not tried this and you should confirm that it will do the right thing:
```
git pull --rebase upstream vNext
git push --force origin
```

ManoShu commented:
@ThomasBarnekow Yes i was using the vNext branch to do the changes.
Once again I am sorry, but while trying to resolve the conflicts to rebase from upstream, my branch is now empty. It is worth it trying to recover it or just re-fork and do the right way? I am assuming that re-forking will make this PR null if I do that?

ThomasBarnekow commented:
@ManoShu I think it does make sense to start from scratch, at least more or less.

I don't know what exactly your local history looks like, but to restart you could either:
* create a completely new fork (as you suggested) or 
* do a hard reset of your local vNext branch to a well-defined commit (that you see in this repo's history), do a `git pull upstream vNext` (where `upstream` is this repository), and lastly do a `git push origin vNext` (where origin is your fork).

You should obviously store away your relevant source code files before doing any of the above.

With a new fork, you would likely have to create a new pull request (on a feature branch, please). That's not a problem, though. You would just close this one and we would continue our discussion on the other one.

EricWhiteDev commented:
Obligatory XKCD reference:

https://xkcd.com/1597/

ManoShu commented:
@EricWhiteDev Yeah, I sure need to git gud, as some say.

@ThomasBarnekow I reforked and redid my changes, however a single test (OxPt.RpTests.RP001(name: "RP/RP052-Deleted-Para-Mark.docx")) failed. As it is related to your Revision Processor changes, I am assuming that I can push my changes, correct?

EricWhiteDev commented:
@ManoShu I struggled (and still struggle) with git, as @ThomasBarnekow can attest to.  I am still ashamed of the commit history that is there in the early parts of the Open-Xml-Sdk.  I made a mess out of it.  Oh well.  Open-Xml-PowerTools is somewhat cleaner, but not perfect.

I can't count the number of times that I have messed up my repo and ended up re-forking / re-cloning, to clean it up.  I am still clueless about squashing commits.

I have just created a PR for fixing the single test that fails.  It will not conflict with your PR - you can create it.

ThomasBarnekow commented:
@ManoShu Yes, as @EricWhiteDev pointed out, he already created a PR to fix the failing unit test, which is totally unrelated to your PR.

I would just want to make sure everything is nice and clean the next time. Thus, I assume you have created a feature branch off of the latest upstream (this repo's) vNext. Right? If so, when you push, your commits to origin (i.e., your fork) you can close this PR and then open a new one.

Now, my next question is about whether or not we need unit tests for this PR. You will obviously want to make sure that some changed output or behavior can be observed, now and in the future. That should be covered by unit tests.

ManoShu commented:
OK, the push was done. Closing this PR.
